elfo-core/src/message/repr.rs

```diff
@@ -0,0 +1,340 @@
+    ) -> fmt::Result {
+        let data = &ptr.cast::<MessageRepr<M>>().as_ref().data;
+        fmt::Debug::fmt(data, f)
+    }
+
+    pub(super) unsafe fn erase<M: Message>(ptr: NonNull<MessageRepr>) -> dumping::ErasedMessage {
+        let data = ptr.cast::<MessageRepr<M>>().as_ref().data.clone();
+        smallbox!(data)
+    }
+
+    /// # Safety
+    ///
+    /// The result pointer is valid only during the lifetime of `ptr`.
+    pub(super) unsafe fn as_serialize_any<M: Message>(
+        ptr: NonNull<MessageRepr>,
+    ) -> NonNull<dyn erased_serde::Serialize> {
+        let data = &ptr.cast::<MessageRepr<M>>().as_ref().data;
+        let ser = data as &dyn erased_serde::Serialize;
+        NonNull::new_unchecked(ser as *const _ as *mut _)
+    }
+
+    pub(super) unsafe fn deserialize_any<M: Message>(
+        deserializer: &mut dyn erased_serde::Deserializer<'_>,
+        out_ptr: NonNull<MessageRepr>,
+    ) -> Result<(), erased_serde::Error> {
+        let data = erased_serde::deserialize::<M>(deserializer)?;
+        ptr::write(
+            out_ptr.cast::<MessageRepr<M>>().as_ptr(),
+            MessageRepr::new(data),
+        );
+        Ok(())
+    }
+
+    cfg_network!({
+        pub(super) unsafe fn read_msgpack<M: Message>(
+            buffer: &[u8],
+            out_ptr: NonNull<MessageRepr>,
+        ) -> Result<(), decode::Error> {
+            let data = decode::from_slice(buffer)?;
+            ptr::write(
+                out_ptr.cast::<MessageRepr<M>>().as_ptr(),
+                MessageRepr::new(data),
+            );
+            Ok(())
+        }
+
+        pub(super) unsafe fn write_msgpack<M: Message>(
+            ptr: NonNull<MessageRepr>,
+            out: &mut Vec<u8>,
+            limit: usize,
+        ) -> Result<(), encode::Error> {
+            let data = &ptr.cast::<MessageRepr<M>>().as_ref().data;
+            let mut out = LimitedWrite(out, limit);
+            encode::write_named(&mut out, data)
+        }
+    });
+}
+
+// === VTable registration & lookup ===
+
+/// A list of all registered message vtables via the `linkme` crate.
+/// Used only for collecting, all lookups are done via a hashmap.
+// Reexported in `elfo::_priv`.
+#[doc(hidden)]
+#[linkme::distributed_slice]
+pub static MESSAGE_VTABLES_LIST: [&'static MessageVTable] = [..];
+
+static MESSAGE_VTABLES_MAP: Lazy<FxHashMap<(&'static str, &'static str), &'static MessageVTable>> =
+    Lazy::new(|| {
+        MESSAGE_VTABLES_LIST
+            .iter()
+            .map(|vtable| ((vtable.protocol, vtable.name), *vtable))
+            .collect()
+    });
+
+impl MessageVTable {
+    /// Finds a vtable by protocol and name.
+    /// Used for deserialization of `AnyMessage` and in networking.
+    pub(crate) fn lookup(protocol: &str, name: &str) -> Option<&'static Self> {
+        // Extend lifetimes to static in order to get `(&'static str, &'static str)`.
+        // SAFETY: this pair doesn't overlive the function.
```

Comment:
That’s mildly scary. Using a custom key type allows to avoid any unsafety here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=04a560d64fba49b3a3a467ea6dd673e4